### DISCUSSION TITLE
How to enable partial redraws for Blender?

### DISCUSSION THREAD

**Discussion by Rush-iam** [Xpra-org/xpra-html5#285] [Q&A]
For some apps, Xpra redraws only updated parts, but for Blender, it always redraws the entire screen, even if I change a small screen value, or hover a button/menu. It makes Blender less responsive, I get a few fps even doing small things.

Is there a way to enable redrawing only updated parts?
I suppose Blender is an OpenGL application, accelerated by VirtualGL.

**Comment by totaam** [comment]
> Is there a way to enable redrawing only updated parts?

My guess is that `blender` is detected as a video application, as such it will tend to use video encodings and those only handle full-window redraws.
Please include the `xpra info` for the window affected.

**Reply by Rush-iam** [reply]
I see `windows.9.content-type=` is empty.
I tried running the server with `XPRA_DEFAULT_CONTENT_TYPE=text` but it made things even worse.
I suppose it uses the `picture` content type, but not sure.

```
windows.9.XShm=True
windows.9.above=
windows.9.allowed-actions=('_NET_WM_ACTION_CLOSE', '_NET_WM_ACTION_MOVE', '_NET_WM_ACTION_RESIZE', '_NET_WM_ACTION_FULLSCREEN', '_NET_WM_ACTION_MINIMIZE', '_NET_WM_ACTION_SHADE', '_NET_WM_ACTION_STICK', '_NET_WM_ACTION_MAXIMIZE_HORZ', '_NET_WM_ACTION_MAXIMIZE_VERT', '_NET_WM_ACTION_CHANGE_DESKTOP', '_NET_WM_ACTION_ABOVE', '_NET_WM_ACTION_BELOW')
windows.9.below=
windows.9.bypass-compositor=
windows.9.children=
windows.9.class-instance=('Blender', 'Blender')
windows.9.client-geometry=(0, 30, 3744, 2066)
windows.9.command=blender
windows.9.content-type=
windows.9.depth=24
windows.9.focused=True
windows.9.fullscreen=
windows.9.geometry=(0, 30, 3744, 2066)
windows.9.grabbed=
windows.9.has-alpha= 
windows.9.icon-title=
windows.9.iconic=
windows.9.maximized=True
windows.9.modal=
windows.9.opacity=-1 
windows.9.opaque-region=
windows.9.override-redirect=
windows.9.pid=21562
windows.9.ppid=17065 
windows.9.protocols=('WM_DELETE_WINDOW', 'WM_TAKE_FOCUS')
windows.9.requested-position=(0, 0)
windows.9.set-initial-position=True
windows.9.shaded=
windows.9.shown=True 
windows.9.size=(3744, 2066)
windows.9.size-constraints.minimum-size=(320, 240)
windows.9.size-constraints.position=(0, 0)
windows.9.skip-pager=
windows.9.skip-taskbar=
windows.9.state=('_NET_WM_STATE_MAXIMIZED_VERT', '_NET_WM_STATE_MAXIMIZED_HORZ', '_NET_WM_STATE_FOCUSED')
windows.9.sticky=
windows.9.title=Blender
windows.9.tray=
windows.9.window-type=('NORMAL',)
windows.9.wm-pid=21562
windows.9.workspace=65535
windows.9.xid=0xa00002
```

**Reply by totaam** [reply]
Without `content-type`, it will use the default logic and guess what is best.

Running your server with `-d compress` will show you which encodings it is using.

**Reply by Rush-iam** [reply]
This is what I get when I hover over a tiny button of Blender:
```
2024-02-22 21:48:16,941 compress:  17.6ms for 3744x2066 pixels at    0,0    for wid=1     using      jpeg with ratio   1.6%  (30216KB to   488KB), sequence    67, client_options={'encoder': 'nvjpeg'}                             , options={'quality': 85, 'speed': 70, 'rgb_formats': ('RGBX', 'RGBA', 'RGB'), 'lz4': True, 'cuda-device-context': cuda_device_context(0)}
```

How to make Xpra redraw only small parts?

**Reply by totaam** [reply]
You would have to make sure that only small parts are actually being updated.
Use `-d damage` for that.

Then the fact that you have `nvjpeg` makes it more likely that it will use it, because it's so fast and efficient. What use the problem that you're seeing with it?

Also, if your version is out of date, please update first, there have been many fixes in this area.

**Reply by Rush-iam** [reply]
> if your version is out of date, please update first

It is impossible for me, I am forced to use Nvidia 470 drivers having CUDA 11.2, and it is not supported in Xpra >=5.0.4 (failing with the missing `BGRX_TO_NV_12` error while initializing).

> Use -d damage for that.
```
2024-02-23 19:45:57,253 damage(0, 0, 3744, 2066, {}) wid=1
2024-02-23 19:45:57,254 do_damage(0, 0, 3744, 2066, {})   wid=1, scheduling batching expiry for sequence   12 in   0 ms
2024-02-23 19:45:57,254  delay=0, elapsed=284, resize_elapsed=8192, congestion_elapsed=20639394, batch=12, min=15, inc=0
2024-02-23 19:45:57,254 expire_delayed_region(20639, 15) delayed region=DelayedRegion(time=20639, expired=False, encoding=auto, regions=[R(0, 0, 3744, 2066)], options={})
2024-02-23 19:45:57,254 do_send_delayed() damage delayed=DelayedRegion(time=20639, expired=True, encoding=auto, regions=[R(0, 0, 3744, 2066)], options={})
2024-02-23 19:45:57,254 acknowledge_changes() xshm handle=XShmWrapper(0x600002 - 3744x2066), damage handle=4194333
2024-02-23 19:45:57,254 invalidating named pixmap, contents handle=PixmapWrapper(0x40002c, 3744, 2066)
```

Oh.. Entire window is damaged, right? Is this because Blender uses OpenGL for its UI?

**Reply by totaam** [reply]
It could be.
All is not lost, `scroll` encoding mode should be able to detect that not much has changed.
Not sure why it didn't fire if that's the case.

**Reply by Rush-iam** [reply]
I remember you disabled `scroll` altogether due to https://github.com/Xpra-org/xpra-html5/issues/231

**Reply by totaam** [reply]
Wait, are you using the html5 client?

**Reply by Rush-iam** [reply]
Yes. This thread is in html-client repo also

**Reply by totaam** [reply]
Oops, I somehow missed that!
You should try with the python client to see if it fares any better.
If it does then we know the html5 client just needs `scroll` implemented.

**Reply by Rush-iam** [reply]
I am restricted to the HTML client only, unfortunately.

But I reverted https://github.com/Xpra-org/xpra-html5/commit/6be1053f5650690634685ee7357351acba68a5b6 locally (it is probably not enough, as I don't get related visual issues) and ran Xpra with `-d scroll`. Here is what I see in the logs:
```
2024-02-26 17:23:24,658 may_use_scrolling(XShmImageWrapper(BGRX: 0, 0, 3744, 2066), {'quality': 75, 'speed': 71, 'rgb_formats': ('RGBX', 'RGBA', 'RGB'), 'lz4': True}) supports_scrolling=False, has_pixels=<built-in method has_pixels of xpra.x11.bindings.ximage.XShmImageWrapper object at 0x7f59e05a1870>, content_type=, non-video encodings=('png', 'webp', 'rgb24', 'rgb32', 'jpeg')
2024-02-26 17:23:24,658 no scrolling: not supported
```

What does it mean?

**Reply by totaam** [reply]
https://github.com/Xpra-org/xpra/blob/5bec0e8a1544fb265ef7f018f235486398777649/xpra/server/window/video_compress.py#L454
Means that at least one end does not support `scroll` encoding.
